### PR TITLE
fix: add type check and fix the type issues

### PR DESCRIPTION
**Pull request by keyvankhademi**

### CODE REVIEW

File: .github/workflows/pre-commit.yaml [hunk 1]
```diff
@@ -10,5 +10,14 @@ jobs:
...
+        with:
+          python-version: '3.10'
+      - name: Run image
+        uses: abatilo/actions-poetry@v2
+        with:
+          poetry-version: '1.3.2'
+      - name: Install dependencies
+        run: poetry install
+      - name: Run pre-commit
+        run: poetry run pre-commit run --all-files
```
Comment (guibescos) [inline .github/workflows/pre-commit.yaml:23]: Why running via poetry instead of via `pre-commit/action@v2.0.3 `

Comment (keyvankhademi) [reply .github/workflows/pre-commit.yaml:23]: The reason behind it is that now we need to have all the packages installed to have their types for the typecheck. And it seems like `pre-commit/action@v2.0.3` is not configurable to work with poetry environment.

File: example_publisher/publisher.py [hunk 2]
```diff
@@ -32,12 +32,20 @@ def __init__(self, config: Config) -> None:
...
-        elif self.config.provider_engine == "pyth_replicator":
+        elif (
+            self.config.provider_engine == "pyth_replicator"
+            and config.pyth_replicator is not None
+        ):
             self.provider: Provider = PythReplicator(config.pyth_replicator)
         else:
-            raise ValueError(f"Unknown provider {self.config.provider_engine}")
+            raise ValueError(
+                f"Unknown provider {self.config.provider_engine}, possibly the env variables is not set."
```
Comment (guibescos) [inline example_publisher/publisher.py:47]: possibly the env variables ~is~ not set

File: example_publisher/pythd.py [hunk 3]
```diff
@@ -15,37 +15,36 @@
...
     metadata: Metadata = field(metadata=config(field_name="attr_dict"))
     prices: List[Price] = field(metadata=config(field_name="price"))
 
 
 class Pythd:
     def __init__(
-        self, address: str, on_notify_price_sched: Callable[[SubscriptionId], Awaitable]
+        self,
+        address: str,
+        on_notify_price_sched: Callable[[SubscriptionId], Coroutine[Any, Any, None]],
```
Comment (guibescos) [inline example_publisher/pythd.py:40]: Why `Any` here, was it hard to have an actual type?

**Review by guibescos — APPROVED** [review]

**Pull request merged**